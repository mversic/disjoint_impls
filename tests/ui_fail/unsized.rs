use disjoint_impls::disjoint_impls;

pub trait Dispatch {
    type Group: ?Sized;
}

pub enum GroupA {}
impl Dispatch for String {
    type Group = GroupA;
}
impl<T> Dispatch for Vec<T> {
    type Group = GroupA;
}

pub struct GroupB;
impl Dispatch for str {
    type Group = GroupB;
}
impl Dispatch for [u8] {
    type Group = GroupB;
}

disjoint_impls! {
    pub trait Kita<T: ?Sized, U: ?Sized> {
        fn kita(&self) -> String;
    }

    impl<T: Dispatch<Group = GroupA>, U: ?Sized, V> Kita<U, V> for T {
        fn kita(&self) -> String {
            "Blanket A".to_owned()
        }
    }

    impl<T: Dispatch<Group = GroupB> + ?Sized, U, V> Kita<U, V> for T {
        fn kita(&self) -> String {
            "Blanket B".to_owned()
        }
    }
}

fn main() {
    assert_eq!("Blanket A", <str as Kita<str, u32>>::kita(&""));
    assert_eq!("Blanket A", <Vec<u8> as Kita<str, str>>::kita(&Vec::new()));
    assert_eq!("Blanket B", <str as Kita<u32, str>>::kita(&""));
    assert_eq!("Blanket B", <[u8] as Kita<str, u32>>::kita(&[]));

    assert_eq!("Blanket A", <String as Kita<str, u32>>::kita(&String::new()));
    assert_eq!("Blanket A", <Vec<u8> as Kita<str, u32>>::kita(&Vec::new()));
    assert_eq!("Blanket B", <str as Kita<u32, u32>>::kita(&""));
    assert_eq!("Blanket B", <[u8] as Kita<u32, u32>>::kita(&[]));
}
