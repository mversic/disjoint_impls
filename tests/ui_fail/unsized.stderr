error[E0277]: the trait bound `str: Kita<str, u32>` is not satisfied
  --> tests/ui_fail/unsized.rs:42:30
   |
42 |     assert_eq!("Blanket A", <str as Kita<str, u32>>::kita(&""));
   |                              ^^^ the trait `Sized` is not implemented for `str`
   |
note: required for `str` to implement `Kita0<GroupB, str, u32>`
  --> tests/ui_fail/unsized.rs:34:54
   |
34 |     impl<T: Dispatch<Group = GroupB> + ?Sized, U, V> Kita<U, V> for T {
   |                                                -     ^^^^^^^^^^     ^
   |                                                |
   |                                                unsatisfied trait bound introduced here
note: required for `str` to implement `Kita<str, u32>`
  --> tests/ui_fail/unsized.rs:28:53
   |
28 |     impl<T: Dispatch<Group = GroupA>, U: ?Sized, V> Kita<U, V> for T {
   |                                                     ^^^^^^^^^^     ^

error[E0277]: the size for values of type `str` cannot be known at compilation time
  --> tests/ui_fail/unsized.rs:43:30
   |
43 |     assert_eq!("Blanket A", <Vec<u8> as Kita<str, str>>::kita(&Vec::new()));
   |                              ^^^^^^^ doesn't have a size known at compile-time
   |
   = help: the trait `Sized` is not implemented for `str`
note: required for `Vec<u8>` to implement `Kita<str, str>`
  --> tests/ui_fail/unsized.rs:28:53
   |
28 |     impl<T: Dispatch<Group = GroupA>, U: ?Sized, V> Kita<U, V> for T {
   |                                                     ^^^^^^^^-^     ^
   |                                                             |
   |                                                             unsatisfied trait bound introduced here

error[E0277]: the trait bound `str: Kita<u32, str>` is not satisfied
  --> tests/ui_fail/unsized.rs:44:30
   |
44 |     assert_eq!("Blanket B", <str as Kita<u32, str>>::kita(&""));
   |                              ^^^ the trait `Sized` is not implemented for `str`
   |
note: required for `str` to implement `Kita<u32, str>`
  --> tests/ui_fail/unsized.rs:28:53
   |
28 |     impl<T: Dispatch<Group = GroupA>, U: ?Sized, V> Kita<U, V> for T {
   |                                                     ^^^^^^^^-^     ^
   |                                                             |
   |                                                             unsatisfied trait bound introduced here

error[E0277]: the size for values of type `str` cannot be known at compilation time
  --> tests/ui_fail/unsized.rs:45:30
   |
45 |     assert_eq!("Blanket B", <[u8] as Kita<str, u32>>::kita(&[]));
   |                              ^^^^ doesn't have a size known at compile-time
   |
   = help: the trait `Sized` is not implemented for `str`
note: required for `[u8]` to implement `Kita0<GroupB, str, u32>`
  --> tests/ui_fail/unsized.rs:34:54
   |
34 |     impl<T: Dispatch<Group = GroupB> + ?Sized, U, V> Kita<U, V> for T {
   |                                                -     ^^^^^^^^^^     ^
   |                                                |
   |                                                unsatisfied trait bound introduced here
note: required for `[u8]` to implement `Kita<str, u32>`
  --> tests/ui_fail/unsized.rs:28:53
   |
28 |     impl<T: Dispatch<Group = GroupA>, U: ?Sized, V> Kita<U, V> for T {
   |                                                     ^^^^^^^^^^     ^
